src/classes/OPP_PrimaryContactRoleMerge_BATCH.cls

```diff
@@ -0,0 +1,127 @@
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2018
+* @group Opportunity Contact Role
+* @group-content ../../ApexDocContent/Rollups.htm
+* @description Batch Class for the deletion of the duplicate primary OCRs.
+*/
+public class OPP_PrimaryContactRoleMerge_BATCH  implements Database.Batchable<sObject> {
+
+    /******************************************************************************************************
+    * @description Batch start method. Runs through all the Opportunities in the org.
+    */
+    public Database.QueryLocator start(Database.batchableContext info) {
+        return Database.getQueryLocator('SELECT Id FROM Opportunity');
+    }
+
+    /*******************************************************************************************************
+    * @description Batch execute method. It gets the Opportunities with duplicate primary OCRs and runs 
+    * the logic that removes the Primary OCR duplicates.
+    * @return void.
+    ********************************************************************************************************/
+    public void execute(Database.BatchableContext BC, List<Opportunity> scope) {
+        List<AggregateResult> primaryOCRsAggregateResult = getPrimaryOCRsAggregateResult(scope);
+        if (primaryOCRsAggregateResult.isEmpty()) {
+            return;
+        }
+
+        OPP_PrimaryContactRoleMerge.DuplicatePrimaryOCRs duplicatePrimaryOCRs  = new OPP_PrimaryContactRoleMerge.DuplicatePrimaryOCRs(primaryOCRsAggregateResult);
+        String settingsRole = UTIL_CustomSettingsFacade.getContactsSettings().npe01__Opportunity_Contact_Role_Default_role__c;
+        validateOpportunityOCRs(duplicatePrimaryOCRs, settingsRole);
```

Comment:
Just commenting about our discussion here, is it possible to make this both readable and efficient so that we are not looping through the same records multiple times?

Reply:
What about placing the validation directly into the merge method or moving the loop higher up so you aren't looping through the same records multiple times doing the same check?